File: haskell/glue/src/Glue/Lib/Bool.hs
module Glue.Lib.Bool where

import Glue.Eval (Eval)
import Glue.IR (IR (..))
import Glue.Lib.Bool.Eq (eq)
import Glue.Lib.Bool.Ge (ge)
import Glue.Lib.Bool.Gt (gt)
import Glue.Lib.Bool.If (if_)
import Glue.Lib.Bool.Le (le)
import Glue.Lib.Bool.Lt (lt)
import Glue.Lib.Bool.Ne (ne)
import Glue.Lib.Bool.Not (not_)
import Glue.Lib.Bool.Until (until_)
import Glue.Lib.Bool.When (when_)
import Glue.Lib.Bool.While (while_)
import Glue.Module (ModuleInfo, nativeModule)

bool :: ModuleInfo Eval
bool =
    nativeModule
        "ffi.bool"
        [ ("true", Bool True)
        , ("false", Bool False)
        , ("==", eq)
        , ("ne", ne)
        , ("!=", ne)
        , ("lt", lt)
        , ("<", lt)
        , ("le", le)
        , ("<=", le)
        , ("gt", gt)
        , (">", gt)
        , ("ge", ge)
        , (">=", ge)
        , ("not", not_)
        , ("!", not_)
        , ("if", if_)
        , ("when", when_)
        , ("while", while_)
        , ("until", until_)
        ]

File: dart/glue/lib/src/lib/bool.dart
import 'package:glue/src/ir.dart';
import 'package:glue/src/lib/bool/and.dart';
import 'package:glue/src/lib/bool/eq.dart';
import 'package:glue/src/lib/bool/fallback.dart';
import 'package:glue/src/lib/bool/ge.dart';
import 'package:glue/src/lib/bool/gt.dart';
import 'package:glue/src/lib/bool/if.dart';
import 'package:glue/src/lib/bool/is_empty.dart';
import 'package:glue/src/lib/bool/is_exist.dart';
import 'package:glue/src/lib/bool/le.dart';
import 'package:glue/src/lib/bool/lt.dart';
import 'package:glue/src/lib/bool/maybe.dart';
import 'package:glue/src/lib/bool/ne.dart';
import 'package:glue/src/lib/bool/not.dart';
import 'package:glue/src/lib/bool/or.dart';
import 'package:glue/src/module.dart';

/// Bool module - boolean operations, comparisons, and control flow
/// Mirrors Haskell Glue.Lib.Bool exactly

/// The bool module containing all boolean functions and special forms
/// Mirrors Haskell Glue.Lib.Bool.bool exactly
final ModuleInfo boolModule = nativeModule('ffi.bool', [
  ('true', IrBool(true)),
  ('false', IrBool(false)),
  ('==', eq),
  ('ne', ne),
  ('!=', ne),
  ('lt', lt),
  ('<', lt),
  ('le', le),
  ('<=', le),
  ('gt', gt),
  ('>', gt),
  ('ge', ge),
  ('>=', ge),

  ('not', not),
  ('!', not),
  ('and', and_),
  ('&&', and_),
  ('or', or_),
  ('||', or_),

  ('is-empty', isEmpty_),
  ('is-exist', isExist_),

  ('fallback', fallback),
  ('??', fallback),

  ('maybe', maybe),
  ('?>', maybe),

  ('if', if_),
  ('?', if_),
]);
